### PR TITLE
fix: use triggering origin for notification links

### LINKED ISSUE #6587: Improve configuration for notification links base urls
As a multi-instance ZITADEL admin, I want to have more control over the base urls in user notifications.
As a single-instance ZITADEL admin, I want the triggers base url to be the default, so that I don't have to care about primary domain configuration.

## Acceptance Criteria

- [ ] Write the notification links base URL into the triggering events. If none is specified in a new request property `notification_base_url`, the requests origin is used.
- [ ] Use `notification_base_url` from events to build notification links instead of the instances primary domain.
- [ ] Fall back to primary domain if `notification_base_url` is not defined in the triggering event from previous ZITADEL versions.
- [ ] Check configurations
- [ ] Update docs, maybe deprecate primary domain

Enables #5395

### PR DESCRIPTION
**Pull request by eliobischof**
Closes #6587, #5844
Enables #5395 

The PR changes which origins are used in notification links.

### Previous behavior

- Assets (Logo): The instances primary domain was used.
- Verification Links: If available in the triggering event, a URL template was parsed and used without further modification. If the template was not enabled, the instances primary domain was used.

### New behavior

- Assets (Logo): If available in the triggering event, the property `triggerOrigin` is used to construct the URL. If not available, the instances primary domain is used.
- Verification Links: If available in the triggering event, a URL template is still parsed and used without further modification. If the template is not available, the property `triggerOrigin` is used to construct the URL. If `triggerOrigin` isn't available neither, the instances primary domain is used to construct the URL. This is for backward compatibility, new events should always have the `triggerOrigin` property set.
- The origin is extracted from the calling context when new events are created.

### Deviations from the AC

- Used the more precise terminology "origin" instead of "base url" 
- Uses the same origin for asset links like for actionable links
- Doesn't change docs

### Other notes

- Refactors the notifications package so that all side effects are mockable.
- It shouldn't be breaking, as the asset URLs must already be servable from the primary domain or the origin domain. 

### Origin source

The origin is added to each request context.
It is constructed from the Forwarded header if available, else from the requests Host header (host and port) and TLS settings (scheme). This aligns the behavior for retrieving information from reverse proxies to the way OIDC issuer information is retrieved. I suggest using the Forwarded header also for discovering virtual instances with #5395 and cleaning up the docs.

### Definition of Ready

- [x] I am happy with the code
- [x] Short description of the feature/issue is added in the pr description
- [x] PR is linked to the corresponding user story
- [x] Acceptance criteria are met
- [x] All open todos and follow ups are defined in a new ticket and justified
- [x] Deviations from the acceptance criteria and design are agreed with the PO and documented.
- [x] No debug or dead code
- [x] My code has no repetitions
- [x] Critical parts are tested automatically
- [x] Where possible E2E tests are implemented
- [x] Documentation/examples are up-to-date
- [x] All non-functional requirements are met
- [ ] Functionality of the acceptance criteria is checked manually on the dev system.

### CODE REVIEW

**Review by muhlemmer — COMMENTED** [review]
PR looks good. Just some minor comments.

**Comment by muhlemmer** [comment]
One more comment:

[Acceptance criteria](https://github.com/zitadel/zitadel/issues/6587) mention:

> Update docs, maybe deprecate primary domain

I didn't see a documentation update in this PR.

File: internal/notification/handlers/gen_mock.go [hunk 1]
```diff
@@ -0,0 +1,4 @@
+package handlers
+
+//go:generate mockgen -package mock -destination ./mock/queries.mock.go github.com/zitadel/zitadel/internal/notification/handlers Queries
+//go:generate mockgen -package mock -destination ./mock/commands.mock.go github.com/zitadel/zitadel/internal/notification/handlers Commands
```
[the remark this answers is not in the document]
Comment (eliobischof) [reply internal/notification/handlers/gen_mock.go:4]: I regenerated :+1:

File: internal/notification/handlers/assets.go [hunk 2]
```diff
@@ -0,0 +1 @@
+package handlers
```
[the remark this answers is not in the document]
Comment (eliobischof) [reply internal/notification/handlers/assets.go:1]: removed :+1:

File: internal/notification/channels.go [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+	registerCounter(c.counters.success.sms, "Successfully delivered SMS")
+	registerCounter(c.counters.failed.sms, "Failed SMS deliveries")
+	registerCounter(c.counters.success.json, "Successfully delivered JSON messages")
+	registerCounter(c.counters.failed.json, "Failed JSON message deliveries")
+	return c
+}
+
+func registerCounter(counter, desc string) {
+	err := metrics.RegisterCounter(counter, desc)
+	logging.WithFields("metric", counter).OnError(err).Panic("unable to register counter")
```
[the remark this answers is not in the document]
Comment (eliobischof) [reply internal/notification/channels.go:62]: Yeah, only during startup.

File: internal/api/authz/context.go [hunk 4]
```diff
@@ -132,6 +132,7 @@ func checkOrigin(ctx context.Context, origins []string) error {
 	if origin == "" {
 		origin = http_util.OriginFromCtx(ctx)
 		if origin == "" {
+			// TODO: Don't we want to enforce an origin header?
```
[the remark this answers is not in the document]
Comment (eliobischof) [reply internal/api/authz/context.go]: :+1:

**Comment by eliobischof** [comment]
I changed the logic for building the origin.

It is constructed from the Forwarded header if available, else from the more widely supported X-Forwarded-* headers, else from the requests Host header (host and port) and TLS settings (scheme).

I suggest using this logic for discovering virtual instances (AC for #5395?) and also for the OIDC issuer (follow-up issue?).
@muhlemmer what do you think?

**Comment by eliobischof** [comment]
> One more comment:
> 
> [Acceptance criteria](https://github.com/zitadel/zitadel/issues/6587) mention:
> 
> > Update docs, maybe deprecate primary domain
> 
> I didn't see a documentation update in this PR.

The changes should be opaque to the users, so I think documenting it would introduce more confusion than clarity. I also looked through the current documentation and I didn't find any text that needs to be changed. @hifabienne is it ok to skip this AC?

**Pull request merged**